### DISCUSSION TITLE
Vercel fails when used with Template Engines

### DISCUSSION THREAD

**Discussion by udofia2** [vercel/community#1194] [Help]
### Summary

Please Can Vercel deploy server side render html using pugjs?

My project uses expressjs with pugjs.

after successful build, it doesn't seem to understand index.pug.

the error output is shown bellow 

### Additional information

2022-12-28T16:22:48.204Z  2022-12-28T16:22:48.271Z      d4fc9a08-93cc-436d-bdb5-21e9572ab32f    ERROR   Error: Failed to lookup view "index" in views directory "src/views"
2022-12-28T16:22:48.204Z      at Function.render (/var/task/node_modules/express/lib/application.js:597:17)
2022-12-28T16:22:48.204Z      at ServerResponse.render (/var/task/node_modules/express/lib/response.js:1039:7)
2022-12-28T16:22:48.204Z      at /var/task/index.js:32:7
2022-12-28T16:22:48.204Z      at Layer.handle [as handle_request] (/var/task/node_modules/express/lib/router/layer.js:95:5)
2022-12-28T16:22:48.204Z      at next (/var/task/node_modules/express/lib/router/route.js:144:13)
2022-12-28T16:22:48.204Z      at Route.dispatch (/var/task/node_modules/express/lib/router/route.js:114:3)
2022-12-28T16:22:48.204Z      at Layer.handle [as handle_request] (/var/task/node_modules/express/lib/router/layer.js:95:5)
2022-12-28T16:22:48.204Z      at /var/task/node_modules/express/lib/router/index.js:284:15
2022-12-28T16:22:48.204Z      at Function.process_params (/var/task/node_modules/express/lib/router/index.js:346:12)
2022-12-28T16:22:48.204Z      at next (/var/task/node_modules/express/lib/router/index.js:280:10)
2022-12-28T16:22:48.204Z  Duration: 26.55 ms    Billed Duration: 27 ms  Memory Size: 1024 MB    Max Memory Used: 104 MB
2022-12-28T16:22:55.082Z  2022-12-28T16:22:55.138Z      68122bf3-4b63-43c8-b949-c65cca14169d    ERROR   Error: Failed to lookup view "index" in views directory "src/views"

### Example

https://odiong-ng.vercel.app/

**Comment by amyegan** [comment]
Hi @udofia2! Can you share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) or more information about your project structure? That would be helpful for collaborative debugging. 

If you prefer to work on it alone, we have a guide about [using Express.js with Vercel](https://vercel.com/guides/using-express-with-vercel) to help with setting up Express in Serverless Functions. It includes a section about the [drawbacks and edge cases](https://vercel.com/guides/using-express-with-vercel#drawbacks-and-edge-cases) to consider as well.

**Reply by udofia2** [reply]
there are two files involved
1.) server.js //This file starts the app
2.) index.pug //This file is the entry point for the frontend

_//server.js_
```
const express = require("express");
const path = require("path");

const app = express();

app.set("view engine", "pug");
app.set("views", "src/views");

app.use(express.static(path.join(__dirname, "src/public")));
app.get("/", (req, res) => {
  res.render("index");
});

```



_//src/views/index.pug_
```doctype html
head
  meta(charset='UTF-8')
  meta(http-equiv='X-UA-Compatible' content='IE=edge')
  meta(name='viewport' content='width=device-width, initial-scale=1.0')
  title template engine
| frontend with template engine(pugjs)
```

//Brief Explanation
It worked with src/views/index.html but not with src/views/index.pug

I will be happy to provide more info where and when needed. Thank you

**Reply by amyegan** [reply]
Does it work if you change `__dirname` to `process.cwd()`?

**Reply by udofia2** [reply]
I made the suggested change but the error is still the same.

Here is the change made
```
app.use(express.static(path.join(process.cwd(), "src/public")));
app.get("/", (req, res) => {
  res.render("index");
});
```

**Reply by amyegan** [reply]
Do you have a public repo or [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) that I can use to debug? The only way I was able reproduce the error was by using `__dirname` which I solved by switching to `process.cwd()`, so there must be something else in your setup that I am missing.

**Reply by udofia2** [reply]
Thank you for taking the time to look into it.

The repo is now publicly available

https://github.com/udofia2/odiong.ng

**Reply by amyegan** [reply]
It looks like your serverless functions are in the root of your application rather than an `/api` directory. You should read through the guide on [using Express.js with Vercel](https://vercel.com/guides/using-express-with-vercel) to help you set up your Express code in a Serverless Function.

**Comment by Crelloc** [comment]
> It looks like your serverless functions are in the root of your application rather than an `/api` directory. You should read through the guide on [using Express.js with Vercel](https://vercel.com/guides/using-express-with-vercel) to help you set up your Express code in a Serverless Function.

My express and pug setup also doesn't work. my express app serves pug files. my webpack build creates a public folder with all the static files, including main.js where, main.js is used inside my base pug file.

I followed the guide you mentioned but i get a 404 error. I'm thinking that vercel is looking for an .html file inside my public folder but none exists because my express app uses pug for templating.

**Reply by Crelloc** [reply]
Also seems like if there's a public folder, vercel assumes that html files are in that directory. I can access static files (from public folder) but doesn't appear that vercel's serverless api is working at all with my express setup of pug server-side rendering

**Comment by Crelloc** [comment]
### Below, I list the necessary files that I have to try to setup vercel serverless functions with express and pug. 

**I can't seem to get this to work. Am i missing anything?**

**Thanks in advance.**

@amyegan 
@timneutkens 
@udofia2 
@sophiekoonin
@leerob 

### Here's my folder structure ... more or less:

```
swsaa-vercel
├── api
│  └── index.js
├── app (folder that has javascript classes and functions)
├── public
├── views
│   ├── pages
│        ├──about.pug
│        ├──collections.pug
│        ├── detail.pug
│        └── home.pug
│   ├── partials
│        ├──navigation.pug
│        └── preloader.pug
│   └──  base.pug
└── app.js (express app)
```

### My package.json file:
```

{
  "name": "swsaa",
  "version": "1.0.0",
  "engines": {
    "node": "18.x"
  },
  "main": "app.js",
    "scripts": {
    "backend:build": "node app.js",
    "backend:development": "nodemon app.js",
    "frontend:build": "webpack --progress --config webpack.config.build.js && cp favicon_io/* public/ && cp sitemap.xml public/",
    "frontend:development": "webpack serve --progress --config webpack.config.development.js",
    "start:con": "concurrently --kill-others \"npm run backend:development\" \"npm run frontend:development\"",
    "start": "node app.js",
    "production": "concurrently --kill-others \"npm run frontend:build\" \"npm run backend:build\"",
    "vercel-build": "echo hello"
  },
  "author": "Thomas Turner",
  "devDependencies": {
    "auto-bind": "^4.0.0",
    "babel-loader": "^8.2.2",
    "clean-webpack-plugin": "^4.0.0-alpha.0",
    "concurrently": "^6.1.0",
    "copy-webpack-plugin": "^8.1.1",
    "css-loader": "^5.2.2",
    "eslint": "^7.24.0",
    "eslint-config-standard": "^16.0.2",
    "eslint-loader": "^4.0.2",
    "eslint-plugin-import": "^2.22.1",
    "eslint-plugin-promise": "^4.3.1",
    "eslint-plugin-standard": "^5.0.0",
    "file-loader": "^6.2.0",
    "glslify-loader": "^2.0.0",
    "gsap": "^3.6.1",
    "image-minimizer-webpack-plugin": "^2.2.0",
    "imagemin": "^7.0.1",
    "imagemin-gifsicle": "^7.0.0",
    "imagemin-jpegtran": "^7.0.0",
    "imagemin-optipng": "^8.0.0",
    "imagemin-svgo": "^9.0.0",
    "include-media": "^1.4.9",
    "mini-css-extract-plugin": "^1.5.0",
    "normalize-wheel": "^1.0.1",
    "ogl": "^0.0.73",
    "postcss-loader": "^5.2.0",
    "prefix": "^1.0.0",
    "raw-loader": "^4.0.2",
    "sass": "^1.32.10",
    "sass-loader": "^11.0.1",
    "terser-webpack-plugin": "^5.1.1",
    "webpack": "^5.33.2",
    "webpack-cli": "^4.6.0",
    "webpack-dev-server": "^3.11.2",
    "webpack-merge": "^5.7.3"
  },
  "dependencies": {
    "@prismicio/client": "^4.0.0",
    "body-parser": "^1.19.0",
    "dotenv": "^8.2.0",
    "errorhandler": "^1.5.1",
    "express": "^4.17.1",
    "helmet": "^4.6.0",
    "method-override": "^3.0.0",
    "morgan": "^1.10.0",
    "nodemon": "^2.0.7",
    "prismic-dom": "^2.2.5",
    "pug": "^3.0.2",
    "ua-parser-js": "^0.7.28"
  },
  "pkg": {
   "assets": [ "./views/**/*", "./public"]
  }
}

```

### The vercel.json files i've tested:

**The following configuration allows me to access everything in the public folder:**
```

{
  "rewrites": [{ "source": "/(.*)", "destination": "/api" }]
}


```

**The following configuration DOESN'T allow me to access everything in the public folder or views folder**

```
{
  "version": 2,
  "builds": [
    {
      "src": "app.js",
      "use": "@vercel/node",
      "config": {
        "includeFiles": ["public/**/*", "views/**/*"]
      }
    }
  ],
  "rewrites": [
    {
      "source": "/(.*)",
      "destination": "/api"
    }
  ]
}

```


### what's inside my api / index.js:

```
import app from '../app.js';

module.exports = app;
```


### Here's my app.js file (express app):


```
require("dotenv").config();
const cluster = require("cluster");
const numCPUs = require("os").cpus().length;
const isDev = process.env.NODE_ENV !== "production";
const helmet = require("helmet");
const { v4 } = require("uuid");
const logger = require("morgan");
const express = require("express");
const errorHandler = require("errorhandler");
const methodOverride = require("method-override");
const find = require("lodash/find");

const app = express();
const path = require("path");
const PORT = 3696;

const Prismic = require("@prismicio/client");
const PrismicDOM = require("prismic-dom");
const UAParser = require("ua-parser-js");

// Multi-process to utilize all CPU cores.
if (!isDev && cluster.isMaster) {
  console.error(`Node cluster master ${process.pid} is running`);

  // Fork workers.
  for (let i = 0; i < numCPUs; i++) {
    cluster.fork();
  }

  cluster.on("exit", (worker, code, signal) => {
    console.error(
      `Node cluster worker ${worker.process.pid} exited: code ${code}, signal ${signal}`
    );
  });
} else {
  app.use(logger("dev"));

  app.use(express.urlencoded({ extended: false }));
  app.use(methodOverride());
  app.use(errorHandler());
  app.use(express.static(path.join(__dirname, "public")));
  app.use(express.static(path.join(__dirname, "views")));

  const initApi = (req) => {
    return Prismic.getApi(process.env.PRISMIC_ENDPOINT, {
      accessToken: process.env.PRISMIC_ACCESS_TOKEN,
      req,
    });
  };

  const handleLinkResolver = (doc) => {
    if (doc.type === "product") {
      return `/detail/${doc.slug}`;
    }

    if (doc.type === "collections") {
      return "/collections";
    }

    if (doc.type === "ab") {
      return "/about";
    }

    return "/";
  };

  app.use((req, res, next) => {
    const ua = UAParser(req.headers["user-agent"]);

    res.locals.isDesktop = ua.device.type === undefined;
    res.locals.isPhone = ua.device.type === "mobile";
    res.locals.isTablet = ua.device.type === "tablet";

    res.locals.Link = handleLinkResolver;

    res.locals.Numbers = (index) => {
      return index == 0
        ? "One"
        : index == 1
        ? "Two"
        : index == 2
        ? "Three"
        : index == 3
        ? "Four"
        : "";
    };

    res.locals.PrismicDOM = PrismicDOM;

    next();
  });

  app.set("views", path.join(__dirname, "views"));
  app.set("view engine", "pug");

  const handleRequest = async (api) => {
    const meta = await api.getSingle("metadata");
    const navigation = await api.getSingle("navigation");
    const preloader = await api.getSingle("preloade");

    return {
      meta,
      navigation,
      preloader,
    };
  };

  app.get("/", async (req, res) => {
    console.log("home path called correctly");

    const api = await initApi(req);
    const defaults = await handleRequest(api);
    const home = await api.getSingle("home");
    const { results: collections } = await api.query(
      Prismic.Predicates.at("document.type", "collec"),
      {
        fetchLinks: "product.image, product.model",
      }
    );

    // console.log("defaults:\n", { ...defaults });
    res.render("pages/home", {
      ...defaults,
      home,
      collections,
    });
  });

  app.get("/about", async (req, res) => {
    const api = await initApi(req);
    const defaults = await handleRequest(api);
    const about = await api.getSingle("ab");
    res.render("pages/about", {
      ...defaults,
      about,
    });
  });

  app.get("/collections", async (req, res) => {
    const api = await initApi(req);
    const defaults = await handleRequest(api);
    const home = await api.getSingle("home");
    const { results: collections } = await api.query(
      Prismic.Predicates.at("document.type", "collec"),
      {
        fetchLinks: "product.image, product.model",
      }
    );
    res.render("pages/collections", {
      ...defaults,
      collections,
      home,
    });
  });

  app.get("/detail/:uid", async (req, res) => {
    const api = await initApi(req);
    const defaults = await handleRequest(api);
    const home = await api.getSingle("home");
    const product = await api.getByUID("product", req.params.uid, {
      fetchLinks: "collec.title",
    });

    console.log(product.data.highlights);
    res.render("pages/detail", {
      ...defaults,
      product,
      home,
    });
  });

  app.post("/__cspreport__", (req, res) => {
    console.log(req.body);
  });

  app.listen(PORT, () => {
    console.log(
      `Node ${
        isDev ? "dev server" : "cluster worker " + process.pid
      }: listening on port ${PORT}`
    );
  });
}

module.exports = app;
```
